Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean

```diff
@@ -0,0 +1,275 @@
+      2⁻¹ ^ n * ∑ ε in ({-1, 1} : Finset ℝ)^^n,
+        ∑ a in A ^^ n, ∑ b in A ^^ n, (∑ i, ε i * (f (a i) - f (b i))) ^ (2 * m) := by
+  let B := A ^^ n
+  have : ∀ ε ∈ ({-1, 1} : Finset ℝ)^^n,
+    ∑ a in B, ∑ b in B, (∑ i, ε i * (f (a i) - f (b i))) ^ (2 * m) =
+      ∑ a in B, ∑ b in B, (∑ i, (f (a i) - f (b i))) ^ (2 * m) :=
+    fun ε hε ↦ step_two_aux A f _ hε fun z : Fin n → ℝ ↦ univ.sum z ^ (2 * m)
+  rw [Finset.sum_congr rfl this, sum_const, card_piFinset_const, card_pair, nsmul_eq_mul,
+    Nat.cast_pow, Nat.cast_two, inv_pow, inv_mul_cancel_left₀]
+  · positivity
+  · norm_num
+
+private lemma step_three (f : α → ℝ) :
+    ∑ ε in ({-1, 1} : Finset ℝ)^^n,
+      ∑ a in A ^^ n, ∑ b in A ^^ n, (∑ i, ε i * (f (a i) - f (b i))) ^ (2 * m) =
+      ∑ a in A ^^ n, ∑ b in A ^^ n, ∑ k in piAntidiag univ (2 * m),
+          (multinomial univ k * ∏ t, (f (a t) - f (b t)) ^ k t) *
+            ∑ ε in ({-1, 1} : Finset ℝ)^^n, ∏ t, ε t ^ k t := by
+  simp only [@sum_comm _ _ (Fin n → ℝ) _ _ (A ^^ n), ← Complex.abs_pow, sum_pow_eq_sum_piAntidiag]
+  refine sum_congr rfl fun a _ ↦ ?_
+  refine sum_congr rfl fun b _ ↦ ?_
+  simp only [mul_pow, prod_mul_distrib, @sum_comm _ _ (Fin n → ℝ), ← mul_sum, ← sum_mul]
+  refine sum_congr rfl fun k _ ↦ ?_
+  rw [← mul_assoc, mul_right_comm]
+
+private lemma step_four {k : Fin n → ℕ} :
+    ∑ ε in ({-1, 1} : Finset ℝ)^^n, ∏ t, ε t ^ k t = 2 ^ n * ite (∀ i, Even (k i)) 1 0 := by
+  have := sum_prod_piFinset ({-1, 1} : Finset ℝ) fun i fi ↦ fi ^ k i
+  rw [this, ← Fintype.prod_boole]
+  have : (2 : ℝ) ^ n = ∏ i : Fin n, 2 := by simp
+  rw [this, ← prod_mul_distrib]
+  refine prod_congr rfl fun t _ ↦ ?_
+  rw [sum_pair, one_pow, mul_boole]
+  swap
+  · norm_num
+  split_ifs with h
+  · rw [h.neg_one_pow, one_add_one_eq_two]
+  rw [← Nat.odd_iff_not_even] at h
+  simp [h.neg_one_pow]
+
+-- double_multinomial
+private lemma step_six {f : α → ℝ} {a b : Fin n → α} :
+    ∑ k : Fin n → ℕ in piAntidiag univ m,
+        (multinomial univ fun a ↦ 2 * k a : ℝ) * ∏ i, (f (a i) - f (b i)) ^ (2 * k i) ≤
+      m ^ m * (∑ i, (f (a i) - f (b i)) ^ 2) ^ m := by
+  rw [sum_pow_eq_sum_piAntidiag, mul_sum]
+  convert sum_le_sum fun k hk ↦ _
+  rw [mem_piAntidiag] at hk
+  simp only [← mul_assoc, pow_mul]
+  gcongr
+  norm_cast
+  refine multinomial_two_mul_le_mul_multinomial.trans ?_
+  rw [hk.1]
+
+private lemma step_seven {f : α → ℝ} {a b : Fin n → α} :
+    m ^ m * (∑ i, (f (a i) - f (b i)) ^ 2 : ℝ) ^ m ≤
+      m ^ m * 2 ^ m * (∑ i, (f (a i) ^ 2 + f (b i) ^ 2)) ^ m := by
+  rw [← mul_pow, ← mul_pow, ← mul_pow, mul_assoc, mul_sum _ _ (2 : ℝ)]
+  gcongr with i
+  exact add_sq_le.trans_eq (by simp)
+
+private lemma step_eight {f : α → ℝ} {a b : Fin n → α} :
+    m ^ m * 2 ^ m * (∑ i, (f (a i) ^ 2 + f (b i) ^ 2) : ℝ) ^ m ≤
+      m ^ m * 2 ^ (m + (m - 1)) *
+        ((∑ i, f (a i) ^ 2) ^ m + (∑ i, f (b i) ^ 2) ^ m) := by
+  rw [pow_add, ← mul_assoc _ _ (2 ^ _), mul_assoc _ (2 ^ (m - 1)), sum_add_distrib]
+  gcongr
```

Comment:
I know this isn't enforced, but I feel pretty strongly that `gcongr` should always have an explicit pattern to make it more robust. At some point I'll make a stronger case for this on zulip, but in the mean time would you mind squeezing the `gcongr`s in this PR?

Reply:
I don't really fancy duplicating all the LHS or RHS every time I use `gcongr`, especially since all `gcongr`s here are maximal (in the sense that the goals it leaves are not of the form `r (f a) (f b)` where `r` is a concrete relation and `f` a concrete operation) so I will wait until you have convinced the rest of the community that this is worth doing

Reply:
Eh, we really should have `gcongr?`